File: Leetcode/tree/865_Smallest Subtree with all the Deepest Nodes/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode subtreeWithAllDeepest(TreeNode root) {
        if(root==null) return null;
        int left = depth(root.left), right = depth(root.right);
        return (left==right)? root: (left > right)?subtreeWithAllDeepest(root.left):
                subtreeWithAllDeepest(root.right);
    }
        private int depth(TreeNode root){
            if(root==null) return 0;
            return 1+Math.max(depth(root.left), depth(root.right));
        }
}

File: Leetcode/tree/865_Smallest Subtree with all the Deepest Nodes/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def subtreeWithAllDeepest(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        if not root:
            return None
        left, right = self.depth(root.left), self.depth(root.right)
        if left==right:
            return root
        return self.subtreeWithAllDeepest(root.left) if left>right else self.subtreeWithAllDeepest(root.right)

    def depth(self, root):
        if not root: return 0
        return 1+max(self.depth(root.left), self.depth(root.right))
